framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,530 @@
+static pthread_mutex_t g_mutex_playing_q;
+static pthread_mutex_t g_mutex_recording_q;
+
+static bool g_playing_live;
+static bool g_recording_live;
+
+static dq_queue_t g_playing_q;
+static dq_queue_t g_recording_q;
+
+/***
+ * Pre-definitions
+ */
+int threadfunc_audio_playing(void *args);
+int threadfunc_audio_recording(void *args);
+
+int media_init(struct media_cb_s *cbs)
+{
+	pthread_attr_t attr;
+	pthread_mutexattr_t mutexAttr;
+	struct sched_param sparam;
+
+	// Init dq and sem
+	dq_init(&g_playing_q);
+	sem_init(&g_sem_playing, 0, 0);
+
+	pthread_mutexattr_init(&mutexAttr);
+	pthread_mutexattr_settype(&mutexAttr, PTHREAD_MUTEX_NORMAL);
+	pthread_mutex_init(&g_mutex_playing_q, &mutexAttr);
+
+
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	dq_init(&g_recording_q);
+	sem_init(&g_sem_recording, 0, 0);
+	pthread_mutex_init(&g_mutex_recording_q, &mutexAttr);
+#endif
+	// Init threads
+	if (pthread_attr_init(&attr) != 0) {
+		return -MEDIA_ERROR_THREAD_INIT;
+	}
+
+	sparam.sched_priority = 150;
+	if (pthread_attr_setschedparam(&attr, &sparam) != 0) {
+		return -MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_attr_setschedpolicy(&attr, SCHED_RR) != 0) {
+		return -MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_attr_setstacksize(&attr, 8192) != 0) {
+		return -MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_create(&g_pth_playing, &attr, (pthread_startroutine_t)threadfunc_audio_playing, (void *)NULL) != 0) {
+		return -MEDIA_ERROR_THREAD_INIT;
+	}
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	if (pthread_create(&g_pth_recording, &attr, (pthread_startroutine_t)threadfunc_audio_recording, (void *)NULL) != 0) {
+		return -MEDIA_ERROR_THREAD_INIT;
+	}
+#endif
+	return MEDIA_OK;
+}
+
+void media_shutdown(void)
+{
+	// Shutdown playback thread
+	g_playing_live = false;
+	sem_post(&g_sem_playing);
+	pthread_join(g_pth_playing, NULL);
+	pthread_mutex_destroy(&g_mutex_playing_q);
+	sem_destroy(&g_sem_playing);
+
+	// Shutdown record thread
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	g_recording_live = false;
+	sem_post(&g_sem_recording);
+	pthread_join(g_pth_recording, NULL);
+	pthread_mutex_destroy(&g_mutex_recording_q);
+	sem_destroy(&g_sem_recording);
+#endif
+}
+
+int media_play(media_t *m, bool loop)
+{
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+	return MEDIA_OK;
+}
+
+int media_record(media_t *m)
+{
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	pthread_mutex_lock(&g_mutex_recording_q);
+	dq_addlast((dq_entry_t *)m, &g_recording_q);
+	pthread_mutex_unlock(&g_mutex_recording_q);
+	sem_post(&g_sem_recording);
+#else
+	pthread_mutex_lock(&g_mutex_playing_q);
+	dq_addlast((dq_entry_t *)m, &g_playing_q);
+	pthread_mutex_unlock(&g_mutex_playing_q);
+	sem_post(&g_sem_playing);
+#endif
+	return MEDIA_OK;
+}
+
+int media_stop(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PLAYING || m->state == MEDIA_STATE_RECORDING) {
+		m->state = MEDIA_STATE_STOPPING;
+		return MEDIA_OK;
+	} else {
```

Comment:
remove else, please

Reply:
I think it is necessary to return error to user if the media object is in some other state than playing or recording. For eg.  you cannot stop already stopped media object.

Reply:
I mean else part just return error.

Reply:
Ok done